include/zephyr/toolchain.h

```diff
@@ -117,15 +117,22 @@
 
 /*
  * Ensure that __BYTE_ORDER__ and related preprocessor definitions are defined,
- * as these are often used without checking for definition and doing so can
- * cause unexpected behaviours.
+ * and that they match the Kconfig option that is used in the code itself to
+ * check for endianness.
  */
 #ifndef _LINKER
 #if !defined(__BYTE_ORDER__) || !defined(__ORDER_BIG_ENDIAN__) || \
     !defined(__ORDER_LITTLE_ENDIAN__)
 
 #error "__BYTE_ORDER__ is not defined"
```

Comment:
just a question here, is this check needed anymore?

Reply:
It's a double-check but I will defer to @tejlmand.

Reply:
yes, as long as we use `__BYTE_ORDER__` and its friends in tree (i.e. right below).

Reply:
Ah right, that's a good reason too :) but we could stop using them in the in-tree toolchain headers entirely if we wanted to (i.e. simplify even further), but I thought that was going too far.

Reply:
If not defined, compiler will also complain about `__BYTE_ORDER__` not defined in the checks below, right? So, the message above doesn't tell me anything the compiler can't.

Reply:
> If not defined, compiler will also complain about `__BYTE_ORDER__` not defined in the checks below, right?

Unless you specify `-Wundef`, no.

Moreover, `-Wundef` is not enabled by default nor is it enabled by `-Wall`.